lib/lte_uart_dfu/lte_uart_dfu.c

```diff
@@ -0,0 +1,340 @@
+		uart_poll_out(context.uart_dev, context.out_buffer[i]);
+	}
+
+	context.out_buf_cursor = 0;
+
+	return 0;
+}
+
+static int lte_uart_dfu_send_error(int func, int error)
+{
+	int err;
+
+	const struct DFU_UART_target_RemoteResp fnc = {
+		.type_choice = func,
+		.error = error
+	};
+
+	err = cbor_encode_DFU_UART_target_RemoteResp(
+		context.out_buffer,
+		ARRAY_SIZE(context.out_buffer),
+		&fnc,
+		&context.out_buf_cursor);
+	if (err) {
+		LOG_ERR("Could not encode response, error %d", err);
+		return err;
+	}
+
+	return lte_uart_dfu_send_out_buffer();
+}
+
+static int lte_uart_dfu_send_offset(int error, size_t offset)
+{
+	int err;
+	int val = error == 0 ? (int) offset : error;
+
+	const struct DFU_UART_target_RemoteResp fnc = {
+		.type_choice = _DFU_UART_target_RemoteFnc_type_offset,
+		.error = val
+	};
+
+	err = cbor_encode_DFU_UART_target_RemoteResp(
+		context.out_buffer,
+		ARRAY_SIZE(context.out_buffer),
+		&fnc,
+		&context.out_buf_cursor);
+	if (err) {
+		LOG_ERR("Could not encode offset, error %d", err);
+		return err;
+	}
+
+	return lte_uart_dfu_send_out_buffer();
+}
+
+static void unused_func(enum dfu_target_evt_id evt)
+{
+	ARG_UNUSED(evt);
+}
+
+static void lte_uart_flush(const struct device *dev)
+{
+	uint8_t c;
+
+	while (uart_fifo_read(dev, &c, 1) > 0) {
+		continue;
+	}
+}
+
+static void uart_cb(const struct device *x, void *p)
+{
+	ARG_UNUSED(p);
+
+	int rx;
+	uint32_t ret;
+	static uint8_t read_buf[128];
+
+	while (uart_irq_update(context.uart_dev) &&
+		uart_irq_rx_ready(context.uart_dev)) {
+		rx = uart_fifo_read(x, read_buf, ARRAY_SIZE(read_buf));
+		if (rx > 0) {
+			ret = ring_buf_put(&context.rx_rb, read_buf, rx);
+			if (ret != rx) {
+				lte_uart_flush(context.uart_dev);
+				k_sem_give(&context.lte_uart_sem);
+				break;
+			}
+			k_sem_give(&context.lte_uart_sem);
+		}
+	}
+}
+
+static int lte_uart_dfu_parse_init(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.file_size = fnc->data.file_size;
+	context.image_num = fnc->data.image_num;
+	context.download_started = false;
+
+	return 0;
+}
+
+static int lte_uart_dfu_parse_write(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.to_write_len = fnc->data.payload.len;
+	context.to_write = fnc->data.payload.value;
+
+	if (!context.download_started) {
+		context.download_started = true;
+	}
+
+	return 0;
+}
+
+static int lte_uart_dfu_parse_done(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.download_started = false;
+	context.download_success = fnc->data.success;
+
+	return 0;
+}
+
+static void execute_input_fnc(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	int err;
+	size_t offs_get;
+	static size_t progress;
+
+	LOG_DBG("DFU function called: %d", fnc->type_choice);
+
+	switch (fnc->type_choice) {
+	case _DFU_UART_target_RemoteFnc_type_init:
+		err = lte_uart_dfu_parse_init(fnc);
+		if (err) {
+			LOG_ERR("Could not parse init, error %d", err);
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_init, err);
+			break;
+		}
```

Comment:
Not possible, `lte_uart_dfu_parse_init` never returns anything other than 0. Why vene bother returning anything then?